src/v/cluster/members_backend.cc

```diff
@@ -569,107 +629,146 @@ void members_backend::calculate_reallocations_batch(
+            // skip partitions outside of current domain
+            if (get_allocation_domain(tp_ns) != domain) {
                 continue;
             }
-        }
-        size_t rand_idx = 0;
-        /**
-         * We use 64 bit field initialized with random number to sample topic
-         * partitions to move, every time the bitfield is exhausted we
-         * reinitialize it and reset counter. This way we chose random
-         * partitions to move without iterating multiple times over the topic
-         * set
-         */
-        std::bitset<sizeof(uint64_t)> sampling_bytes;
-        for (const auto& p : metadata.get_assignments()) {
-            if (idx_it->second > current_idx++) {
+            // do not try to move internal partitions
+            if (
+              tp_ns.ns == model::kafka_internal_namespace
+              || tp_ns.ns == model::redpanda_ns) {
                 continue;
             }
-
-            if (rand_idx % sizeof(uint64_t) == 0) {
-                sampling_bytes = random_generators::get_int(
-                  std::numeric_limits<uint64_t>::max());
-                rand_idx = 0;
-            }
-
-            idx_it->second++;
-
-            if (sampling_bytes.test(rand_idx++)) {
+            if (!metadata.is_topic_replicable()) {
                 continue;
             }
+            // do not move topics that were created after node was added, they
+            // are allocated with new cluster capacity
+            if (meta.update) {
+                if (metadata.get_revision() > meta.update->offset()) {
+                    vlog(
+                      clusterlog.debug,
+                      "skipping reallocating topic {}, its revision {} is "
+                      "greater than node update {}",
+                      tp_ns,
+                      metadata.get_revision(),
+                      meta.update->offset);
+                    continue;
+                }
+            }
 
-            std::erase_if(to_move_from_node, [](const replicas_to_move& v) {
-                return v.left_to_move == 0;
-            });
-
-            std::sort(to_move_from_node.begin(), to_move_from_node.end(), cmp);
-            for (auto& to_move : to_move_from_node) {
-                if (
-                  is_in_replica_set(p.replicas, to_move.id)
-                  && to_move.left_to_move > 0) {
+            for (const auto& p : metadata.get_assignments()) {
+                if (is_in_replica_set(p.replicas, to_move.id)) {
                     partition_reallocation reallocation(
                       model::ntp(tp_ns.ns, tp_ns.tp, p.id), p.replicas.size());
                     reallocation.replicas_to_remove.emplace(to_move.id);
                     auto current_assignment = p;
-                    reassign_replicas(current_assignment, reallocation);
-                    // skip if partition was reassigned to the same node
-                    if (is_reassigned_to_node(reallocation, to_move.id)) {
```

Comment:
I think this check is still relevant, even if reallocation is the same, the error can improve a tiny bit due to rounding errors